zk/sequencer/transaction_yielder.go

```diff
@@ -0,0 +1,350 @@
+	ctx context.Context,
+	cfg ethconfig.Zk,
+	pool *txpool.TxPool,
+	yieldSize uint16,
+	db kv.RwDB,
+	decodedTxCache *expirable.LRU[common.Hash, *types.Transaction],
+) *PoolTransactionYielder {
+	// Initialize the channel with the specified size
+	readyTransactions := make([]common.Hash, 0)
+
+	return &PoolTransactionYielder{
+		readyTransactions:     readyTransactions,
+		readyTransactionBytes: make(map[common.Hash][]byte),
+		readyMtx:              sync.Mutex{},
+		toSkip:                make(map[common.Hash]struct{}),
+		ctx:                   ctx,
+		cfg:                   cfg,
+		pool:                  pool,
+		yieldSize:             yieldSize,
+		db:                    db,
+		decodedTxCache:        decodedTxCache,
+		startedYielding:       false,
+		startedYieldingMtx:    sync.Mutex{},
+	}
+}
+
+func (y *PoolTransactionYielder) YieldNextTransaction() (types.Transaction, uint8, bool) {
+	var tx types.Transaction
+	var effectiveGas uint8
+	var err error
+	var yieldedSomething bool
+
+	y.readyMtx.Lock()
+	defer y.readyMtx.Unlock()
+
+	for _, hash := range y.readyTransactions {
+		if _, found := y.toSkip[hash]; found {
+			continue
+		}
+		if txBytes, found := y.readyTransactionBytes[hash]; found {
+			txPtr, inCache := y.decodedTxCache.Get(hash)
+			if inCache {
+				tx = *txPtr
+			} else {
+				tx, err = types.DecodeTransaction(txBytes)
+				if err != nil {
+					log.Warn("[extractTransaction] Failed to decode transaction from ready queue, skipping and removing from queue",
+						"error", err,
+						"id", hash.String())
+					y.pool.MarkForDiscardFromPendingBest(hash)
+					y.toSkip[hash] = struct{}{}
+					continue
+				}
+				y.decodedTxCache.Add(hash, &tx)
+			}
+			effectiveGas = deriveEffectiveGasPrice(y.cfg, tx)
+			yieldedSomething = true
+			break
+		}
+	}
+
+	return tx, effectiveGas, yieldedSomething
+}
+
+func (y *PoolTransactionYielder) RemoveMinedTransactions(hashes []common.Hash) {
+	y.readyMtx.Lock()
+	defer y.readyMtx.Unlock()
+
+	for _, hash := range hashes {
+		y.decodedTxCache.Remove(hash)
+	}
+
+	// ensure we take a fresh view on the pool
+	y.toSkip = make(map[common.Hash]struct{})
+}
+
+func (y *PoolTransactionYielder) AddMined(hash common.Hash) {
+	y.readyMtx.Lock()
+	defer y.readyMtx.Unlock()
+	y.toSkip[hash] = struct{}{}
+
+	// remove the transaction from the readyTransactions slice. this will save burning CPU for the next yielding
+	// for a transaction to execute.  we still maintain the hash in the toSkip map to avoid yielding it again
+	// when we refresh the pool best list into the readyTransactions slice. there is a window where we have
+	// executed something, but the pool hasn't removed it yet, so we could yield it again before this has happened
+	foundIdx := -1
+	for idx, readyHash := range y.readyTransactions {
+		if readyHash == hash {
+			foundIdx = idx
+			break
+		}
+	}
+	if foundIdx != -1 {
+		y.readyTransactions = append(y.readyTransactions[:foundIdx], y.readyTransactions[foundIdx+1:]...)
```

Comment:
I think here we could speed up here by using swap and pop trick - that will decrease amount of iterations for searching the hash as the latter elements will be found faster as they will be in the beginning, and no elements copy will happen under the hood:

```
last := len(y.readyTransactions) - 1
y.readyTransactions[foundIdx] = y.readyTransactions[last]
y.readyTransactions = y.readyTransactions[:last]
```
that gives some 12% performance improvement for this operation in my benchmark of 10-100K slice size.

Reply:
This is an interesting suggestion and I can certainly see it would be faster, just moving 1 thing and re-slicing the size.... but I worry about ordering, if I take the last element and move it to the middle of the slice I've effectively moved a bad transaction up the queue to be processed before "better" transactions.

Reply:
Ah, that is true. I somehow thought since we pass hash in here, we'll be doing just fine. But we peek the first item from slice (iterate over the readyTransactions) on yielding. So in case we want this improvement, we ll need to keep an index of the last fetched tx. then we proceed yielding since that position. I think that is easy. I could do that in one of the follow ups